Read remote SQLite databases through a DuckDB FileSystem VFS

ak2k opened the pull request:
Reading a SQLite database that lives in object storage or at a URL currently means downloading it first. This lets `sqlite_scan` and `ATTACH` open it in place:

```sql
SELECT * FROM sqlite_scan('https://example.com/chinook.sqlite', 'Artist');

ATTACH 'https://example.com/chinook.sqlite' AS db (TYPE sqlite);
SELECT count(*) FROM db.Album;
```

A remote path is opened through a SQLite VFS that forwards reads to DuckDB's `CachingFileSystem`; local paths keep the native SQLite code path. Because it sits at the FileSystem layer, it works for any filesystem DuckDB exposes (HTTP, S3, GCS, Azure, …) and reuses DuckDB's existing file cache. Queries fetch byte ranges on demand rather than downloading the whole file.

**Limitations**
- Remote databases are read-only. As a consequence, SQLite's temp store stays in memory, so a large sort or hash that would normally spill to disk is kept in RAM.
- The database must be checkpointed and self-contained; reading a remote WAL is not supported. `PRAGMA wal_checkpoint(TRUNCATE)` or `VACUUM INTO` produces a servable copy.
- A query treats the remote object as immutable (the VFS reports `SQLITE_IOCAP_IMMUTABLE`). httpfs raises an error if the object's ETag changes mid-read; on a server that returns no stable ETag, a concurrent change can be read inconsistently.
- Some network-error classification matches on httpfs message text, so it can be imprecise if those messages change.

Tests are hermetic (after setup): a stdlib Python HTTP server under `scripts/` serves committed fixtures over localhost, so core CI needs no network. Optional S3 and high-concurrency paths are gated on environment flags and use a pinned `rclone` binary the CI fetches. The remote tests pull `httpfs` (pinned to the commit the bundled DuckDB engine coordinates, with its patches) as a test-only dependency; the WASM build skips it — those tests don't run there, and httpfs's OpenSSL dependency didn't build for emscripten in our build. The extension compiles for the WASM targets, where all paths route through the FileSystem; that path is not exercised at runtime.

**Notes**
- The remote connection opens lazily, per transaction, on first use. A catalog-level open (one per attachment) is also feasible — the file cache is reachable from the database instance — matching the approach suggested in #183/#184; happy to switch to that if you'd prefer.
- Block data is served from DuckDB's file cache after the first read; the `-wal`/`-journal` sidecar checks are not — each open re-probes them with a network `HEAD`, so repeated and parallel queries re-pay those round-trips.

Related Issues: #39, #141

Maxxen reviewed (commented):
Hello!

Looks cool, but I wonder why/if this vfs is limited to http, why not just wrap DuckDBs regular filesystem? You can then let DuckDB handle the underlying abstraction and ensure sqlite can access files through all duckdb supported filesystems (e.g. compressed files, or regular files on WASM)

ak2k committed:
Add ClientContext parameter to SQLiteDB::Open for future extensibility

Also adds yum support for sqlite3 installation in CI.

ak2k committed:
Fix potential deadlock when opening remote SQLite databases

Opening SQLite connections for remote files can trigger HTTP requests while
the MetaTransaction lock is held, potentially causing deadlocks. This change
defers connection initialization until first use, with thread-safe lazy
initialization using atomic flags and per-database mutexes.

ak2k committed:
Add SQLite VFS implementation for remote file support

ak2k committed:
Enable HTTP/HTTPS SQLite database access via sqlite_scan

This commit adds support for querying remote SQLite databases over HTTP/HTTPS
using the sqlite_scan() function. Remote files are accessed through DuckDB's
custom VFS with caching support.

Implementation details:
- Add OpenWithVFS() to handle remote SQLite databases via custom VFS
- Implement helper methods for clean separation of local vs remote handling
- Register VFS cleanup callback to properly clean up on connection close
- Add comprehensive error handling with HTTP status code mapping
- Pass ClientContext through scanner to enable VFS functionality

The implementation uses DuckDB's CachingFileSystem for efficient block-level
caching of remote file data, minimizing network requests when accessing
SQLite databases over HTTP.

ak2k committed:
Add HTTP ATTACH support and improve error handling

This commit completes the HTTP SQLite support by enabling ATTACH functionality
for remote databases and improving error handling throughout the codebase.

Key changes:
- Replace generic std::runtime_error with specific DuckDB exception types
  (BinderException, ConnectionException, IOException, InternalException)
- Update GetSQLiteTransaction to handle missing transactions gracefully
- Add validation for busy_timeout in SQLiteAttach to prevent overflow
- Implement proper move semantics using swap
- Add Copy/Equals methods to AttachFunctionData for proper state management
- Enable ATTACH functionality for remote SQLite databases
- Defer transaction start to prevent deadlocks with remote file access

FIGIO55 commented:
Trying to see if there are updates on this matter since it would be very useful for me

ak2k committed:
Add a DuckDB FileSystem-backed SQLite VFS for remote files

A custom SQLite VFS that delegates file I/O to DuckDB's CachingFileSystem, so any
filesystem DuckDB exposes (HTTP, S3, GCS, Azure, etc.) is reachable as a read-only
SQLite database. Registered per ClientContext and unregistered on connection close.

ak2k committed:
Add hermetic remote-SQLite test infrastructure, tests, and CI

A stdlib HTTP range server (and optional rclone http/s3 backends) serves committed
SQLite fixtures over localhost, so the remote tests run hermetically. Adds the
http_sqlite_* suite and a CI workflow that runs it under sanitizers.

ak2k committed:
Route remote paths through the VFS in sqlite_scan and ATTACH

SQLiteDB::Open now takes a ClientContext and routes paths DuckDB's FileSystem owns
(remote, WASM) through the caching VFS, opened read-only; local files keep the
native SQLite path unchanged. Open errors from the VFS path are enriched with the
underlying filesystem error (HTTP status/URL).

ak2k committed:
Prepare the transaction layer for remote SQLite databases

Defer the connection open and BEGIN for remote (HTTP/S3) databases to first use
(GetDB()), so attaching a remote database does not open it eagerly. Destroying a
SQLiteTransaction runs sqlite3_close_v2, which can re-enter paths that take
transaction_lock; extract the transaction under the lock and let it destruct
after the lock is released to avoid a lock-order inversion.

ak2k committed:
Bump duckdb submodule to fix external file cache concurrency race

The concurrent-scans test first-accesses a remote file from several
connections at once, tripping a data race in DuckDB's external file cache.
Upstream duckdb#22979 fixes it by requesting parallel access on the shared
file handle. Advance the submodule to the merge commit that carries the fix.

ak2k committed:
Run the concurrent-scans test against the rclone-backed server

http_sqlite_02 opens several remote connections at once. CPython's stdlib
HTTP server does not serve simultaneous connections reliably, so the test
intermittently fails (~1% disk I/O errors) on the default server while
passing cleanly under rclone serve http. Gate it on SQLITE_HTTP_ROBUST like
the concurrent-lifecycle test, and widen the rclone CI steps to cover both.

staticlibs commented on src/storage/sqlite_transaction.cpp:
```diff
@@ -52,15 +53,20 @@ void SQLiteCatalogMap::EraseEntry(const string &entry_name) {
-		owned_db = SQLiteDB::Open(sqlite_catalog.path, sqlite_catalog.options, true);
+		initialized.store(true, std::memory_order_release);
+	} else if (!FileSystem::IsRemoteFile(sqlite_catalog.path)) {
+		// local on-disk database - open eagerly (cheap; surfaces open errors at attach time)
```
M, this comment is technically incorrect, we actually now have 3 places where SQLite DB is opened:

 - `:memory:` DBs - on `ATTACH` call, in `SQLiteCatalog` constructor
 - local files - when transaction is created on catalog access after the `ATTACH`
 - remote files - on first `GetDB`

Don't suggest to change it now, just I think this needs to be eventually made better structured.

staticlibs commented on src/sqlite_duckdb_vfs_cache.cpp:
```diff
@@ -0,0 +1,963 @@
+	// Group 1: "status_code":"XXX" (JSON)
+	// Group 2: XXX (Description) (httpfs format)
+	// Group 3: (HTTP XXX) or HTTP code XXX or HTTP XXX
+	static duckdb_re2::Regex status_regex(
```
It may be better to switch this to `std::regex` - at some point in future we want to run the scanner on top of C API, so don't want unnecessary deps.

staticlibs commented on src/sqlite_duckdb_vfs_cache.cpp:
```diff
@@ -0,0 +1,963 @@
+					    "\"-wal\" sidecar failed (%s). Refusing to read rather than risk serving a stale "
+					    "snapshot. Checkpoint the database into a single file (PRAGMA "
+					    "wal_checkpoint(TRUNCATE), or VACUUM INTO), or make the \"-wal\" URL reachable.",
+					    path, ErrorData(e).RawMessage());
```
This logic looks very thorough, still I think it may make sense to add an option allowing to skip `-journal` probe. We are running read-only after all and I can imagine situation when `-journal` URL won't be accessible.

staticlibs commented on src/sqlite_duckdb_vfs_cache.cpp:
```diff
@@ -0,0 +1,963 @@
+//===--------------------------------------------------------------------===//
+
+// System-level ops (randomness, sleep, time) have no remote file I/O; delegate to the default VFS.
+#define DELEGATE_TO_DEFAULT_VFS(method_name, ...)                                                                      \
```
I think it would be great to convert this macro into some kind of a template helper, if possible.

staticlibs commented on src/sqlite_duckdb_vfs_cache.cpp:
```diff
@@ -0,0 +1,963 @@
+
+		    ClientContext *context = wrapper->context;
+
+		    if (!context || !context->db) {
```
I guess this check is "best effort" currently, as we cannot really check the context liveness in a safe way here.

staticlibs reviewed (approved):
Thanks again for the PR and sorry for the delay! Overall I think this is a robust implementation and want to praise the thorough approach to concurrency (especially in shutdown).

I added a few comments, but they are minor and not all actionable. The ones that are actionable can be addressed in future.

I will try to merge this as-is now, if I get troubles with merging (GH seems to not like the submodule update that I myself suggested) - will rebase the commits (excluding submodule update) onto the main and merge as a new PR.
